File: evals/retrievers/rerank.ts
import type { RerankRow } from '../embed/rerank-precompute.ts'
import type { Retriever } from './keyword.ts'

/**
 * Serves the precomputed cross-encoder ordering. Fails closed on a query it has no row for rather
 * than returning nothing: an empty result would score as a miss and read as the reranker being
 * bad, when the real cause is a stale precompute.
 */
export function rerankRetriever(rows: RerankRow[], name = 'rerank'): Retriever {
  const byQuery = new Map(rows.map((r) => [r.query, r.ranked]))
  return {
    name,
    search(query, k) {
      const ranked = byQuery.get(query)
      if (!ranked) {
        throw new Error(`no reranked order for query: ${query.slice(0, 60)} — rerun eval:rerank`)
      }
      return ranked.slice(0, k)
    },
  }
}

File: evals/retrievers/rerank.py
from typing import TypedDict


class RerankRow(TypedDict):
    query: str
    candidateHash: str
    ranked: list[str]
    scores: list[float]
    depth: int
    maxChars: int


class RerankRetriever:
    def __init__(self, rows: list[RerankRow], name: str = 'rerank') -> None:
        self.name = name
        self.by_query = {r['query']: r['ranked'] for r in rows}

    def search(self, query: str, k: int) -> list[str]:
        ranked = self.by_query.get(query)
        if ranked is None:
            raise LookupError(f'no reranked order for query: {query[:60]} — rerun eval:rerank')
        return ranked[:k]
